src/hydro/srcterms/tabular_cooling.cpp

```diff
@@ -519,7 +579,130 @@ void TabularCooling::MixedIntSrcTerm(parthenon::MeshData<parthenon::Real> *md,
+          // ConservedToPrim conversion, but keeping it for now (better safe than sorry).
+          prim(IPR, k, j, i) = rho * internal_e_floor * gm1;
+          return;
+        }
+
+        auto temp = mu_m_u_gm1_by_k_B * internal_e;
+        // Temperature is above floor (see conditional above) but below cooling table:
+        // -> no cooling
+        if (temp < temp_cool_floor) {
+          return;
+        }
+        const Real n_h2_by_rho = rho * X_by_m_u * X_by_m_u;
+
+        // Get the index of the right temperature bin
+        // TODO(?) this could be optimized for using a binary search
+        auto idx = 0;
+        while ((idx < nbins - 1) && (temps(idx + 1) < temp)) {
+          idx += 1;
+        }
+
+        // Compute the Temporal Evolution Function Y(T) (Eq. A5)
+        const auto alpha_k_m1 = alpha_k(idx) - 1.0;
+        const auto tef =
+            Y_k(idx) + (lambda_final / lambdas(idx)) * (temps(idx) / temp_final) *
+                           (std::pow(temps(idx) / temp, alpha_k_m1) - 1.0) / alpha_k_m1;
+
+        // Compute the adjusted TEF for new timestep (Eqn. 26) (term in brackets)
+        const auto tef_adj =
+            tef + lambda_final * dt / temp_final * mu_m_u_gm1_by_k_B * n_h2_by_rho;
+
+        // TEF is a strictly decreasing function and new_tef > tef
+        // Check if the new TEF falls into a lower bin, i.e., find the right bin for A7
+        // If so, update slopes and coefficients
+        while ((idx > 0) && (tef_adj > Y_k(idx))) {
+          idx -= 1;
+        }
+
+        // Compute the Inverse Temporal Evolution Function Y^{-1}(Y) (Eq. A7)
+        const auto temp_new =
+            temps(idx) *
+            std::pow(1 - (1.0 - alpha_k(idx)) * (lambdas(idx) / lambda_final) *
+                             (temp_final / temps(idx)) * (tef_adj - Y_k(idx)),
+                     1.0 / (1.0 - alpha_k(idx)));
+        // Set new temp (at the lowest to the lower end of the cooling table)
+        const auto internal_e_new = temp_new > temp_cool_floor
+                                        ? temp_new / mu_m_u_gm1_by_k_B
+                                        : temp_cool_floor / mu_m_u_gm1_by_k_B;
+        cons(IEN, k, j, i) += rho * (internal_e_new - internal_e);
+        // Latter technically not required if no other tasks follows before
+        // ConservedToPrim conversion, but keeping it for now (better safe than sorry).
+        prim(IPR, k, j, i) = rho * internal_e_new * gm1;
+      });
+}
+
 Real TabularCooling::EstimateTimeStep(MeshData<Real> *md) const {
+  // No need to restrict dt for townsend cooling
+  // TODO(pgrete) make this optional so that the cfl_cool is detached
+  // from the cooling mechanism, because it may still be desireable to
+  // not let different physical processes evolve on vastly different
+  // timescales.
+  if (integrator_ == CoolIntegrator::townsend) {
+    return std::numeric_limits<Real>::max();
+  }
```

Comment:
This is why there is a CFL condition for the other cooling methods as well -- we didn't want the gas to cool over a single hydrodynamic time step. The adaptive integrators should be able to integrate the cooling over any timestep just fine. Alternatively, since the gas does cool very fast, having a window on how much the cooling time can limit the hydro step might be helpful.

For example, limit the hydro timestep to less than 0.1 of the cooling time but not less than 0.01 of the original hydro timestep. Does that make sense?

Reply:
There is a already a parameter for this: `cooling/cfl` which behaves exactly as you say. It's not tied to the integration of the source term, it just needs to be able to compute `DeDt` with the in kernel in `EstimateTimeStep` everywhere to find the meshblock minimum (ie. max cooling in the meshblock). If your cooling table is evenly spaced (which it's not) and you're ok with the linear interpolation in log space for the purposes of timestep estimation, then `DeDt` will work for any integrator.

Is the larger issue that `DeDt` needs to work for different cooling tables (not just even spaced, 0 cooling below and free-free above) and the cooling defined for the Townsend integrator?